pkg/erofs/erofs.go

```diff
@@ -0,0 +1,691 @@
+	FEATURE_INCOMPAT_XATTR_PREFIXES = 0x00000040
+
+	FEATURE_INCOMPAT_ALL = 0x0
+)
+
+// File types used in Dirent::FileType.
+const (
+	FT_UNKNOWN = iota
+	FT_REG_FILE
+	FT_DIR
+	FT_CHRDEV
+	FT_BLKDEV
+	FT_FIFO
+	FT_SOCK
+	FT_SYMLINK
+	FT_MAX
+)
+
+// Sizes of the on-disk structures in bytes.
+const (
+	SUPER_BLOCK_SIZE    = 128
+	INODE_COMPACT_SIZE  = 32
+	INODE_EXTENDED_SIZE = 64
+	DIRENT_SIZE         = 12
+)
+
+// SuperBlock represents on-disk super block for EROFS:
+// https://docs.kernel.org/filesystems/erofs.html
+//
+// +marshal
+// +stateify savable
+type SuperBlock struct {
+	Magic              uint32
+	Checksum           uint32
+	FeatureCompat      uint32
+	BlockSizeBits      uint8
+	SuperBlockExtSlots uint8
+	RootNid            uint16
+	Inodes             uint64
+	BuildTime          uint64
+	BuildTimeNsec      uint32
+	Blocks             uint32
+	MetaBlockAddr      uint32
+	XattrBlockAddr     uint32
+	UUID               [16]uint8
+	VolumeName         [16]uint8
+	FeatureIncompat    uint32
+	Union1             uint16
+	ExtraDevices       uint16
+	DevTableSlotOff    uint16
+	Reserved           [38]uint8
+}
+
+// BlockSize returns the block size.
+func (sb *SuperBlock) BlockSize() uint32 {
+	return 1 << sb.BlockSizeBits
+}
+
+// BlockAddrToOffset converts block addr to the offset in image file.
+func (sb *SuperBlock) BlockAddrToOffset(addr uint32) uint64 {
+	return uint64(addr) << sb.BlockSizeBits
+}
+
+// MetaOffset returns the offset of metadata area in image file.
+func (sb *SuperBlock) MetaOffset() uint64 {
+	return sb.BlockAddrToOffset(sb.MetaBlockAddr)
+}
+
+// NidToOffset converts inode number to the offset in image file.
+func (sb *SuperBlock) NidToOffset(nid uint64) uint64 {
+	return sb.MetaOffset() + (nid << ISLOT_BITS)
+}
+
+// InodeCompact represents 32-byte reduced form of on-disk inode.
+//
+// +marshal
+type InodeCompact struct {
+	Format       uint16
+	XattrCount   uint16
+	Mode         uint16
+	Nlink        uint16
+	Size         uint32
+	Reserved     uint32
+	RawBlockAddr uint32
+	Ino          uint32
+	UID          uint16
+	GID          uint16
+	Reserved2    uint32
+}
+
+// InodeExtended represents 64-byte complete form of on-disk inode.
+//
+// +marshal
+type InodeExtended struct {
+	Format       uint16
+	XattrCount   uint16
+	Mode         uint16
+	Reserved     uint16
+	Size         uint64
+	RawBlockAddr uint32
+	Ino          uint32
+	UID          uint32
+	GID          uint32
+	Mtime        uint64
+	MtimeNsec    uint32
+	Nlink        uint32
+	Reserved2    [16]uint8
+}
+
+// Dirent represents on-disk directory entry.
+//
+// The size of this struct is bigger than the marshalled size due to
+// the implicit padding at the end of the struct.
+//
+// +marshal
+type Dirent struct {
+	Nid      uint64
+	NameOff  uint16
+	FileType uint8
+	Reserved uint8 `marshal:"unaligned"`
+}
+
+// Image represents an open EROFS image.
+//
+// +stateify savable
+type Image struct {
+	src   *os.File `state:"nosave"`
```

Comment:
We can’t remove `src`, since it will be used in `func (i *Inode) FD() int` to return the host FD of the underlaying image file, which will be used in e.g. `MapInternal()` in `pkg/sentry/fsimpl/erofs/regular_file.go` to handle the page faults on the mappings on the files on EROFS.

Reply:
We can’t remove `src`, since it will be used in `func (i *Inode) FD() int` to return the host FD of the underlaying image file, which will be used in e.g. `MapInternal()` in `pkg/sentry/fsimpl/erofs/regular_file.go` to handle the page faults on the mappings on the files on EROFS.